Development/nmos/ocsp_behaviour.cpp

```diff
@@ -0,0 +1,362 @@
+
+            // start a background task to continously request certificate status on a given interval
+            return pplx::do_while([=, &model, &ocsp_state, &state, &gate]
+            {
+                auto request_interval = std::chrono::seconds(0);
+                if (state.base_uri == state.client->base_uri())
+                {
+                    auto interval = std::uniform_int_distribution<>(
+                        state.next_request > 0.0 ? std::min((int)state.next_request, ocsp_interval_min) : ocsp_interval_min,
+                        state.next_request > 0.0 ? std::min((int)state.next_request, ocsp_interval_max) : ocsp_interval_max)(state.engine);
+                    request_interval = std::chrono::seconds(interval);
+
+                    slog::log<slog::severities::more_info>(gate, SLOG_FLF) << "Waiting to request certificate status for about " << interval << " seconds";
+                }
+                else
+                {
+                    state.base_uri = state.client->base_uri();
+                }
+
+                auto time_now = std::chrono::steady_clock::now();
+                return pplx::complete_at(time_now + request_interval, token).then([=, &state, &gate]()
+                {
+                    return request_certificate_status(*state.client, state.ocsp_request, gate, token);
+                }).then([&model, &ocsp_state, &state, &gate](std::vector<uint8_t> ocsp_response)
+                {
+                    // cache the OCSP response
+                    slog::log<slog::severities::info>(gate, SLOG_FLF) << "Cache the OCSP response";
+                    nmos::with_write_lock(ocsp_state.mutex, [&] { ocsp_state.ocsp_response = ocsp_response; });
+
+                    return true;
+                });
+            }).then([&](pplx::task<void> finally)
+            {
+                auto lock = model.write_lock(); // in order to update local state
+
+                try
+                {
+                    finally.get();
+                }
+                catch (const web::http::http_exception& e)
+                {
+                    slog::log<slog::severities::error>(gate, SLOG_FLF) << "Certificate status request HTTP error: " << e.what() << " [" << e.error_code() << "]";
+                }
+                catch (const ocsp_service_exception&)
+                {
+                    slog::log<slog::severities::error>(gate, SLOG_FLF) << "Certificate status request error";
+                }
+
+                // reaching here, there must be something has gone wrong with the OCSP server
+                // let's select the next available OCSP server
+                state.ocsp_service_error = true;
+
+                model.notify();
+            });
+        }
+
+        double half_certificate_expiry_from_now(const std::vector<utility::string_t>& certificate_chains, slog::base_gate& gate)
+        {
+            double expiry_time = -1.0;
```

Comment:
Wouldn't this all be much simpler if you initialized this, and `next_request` to `(std::numeric_limits<double>::max)()`?

Reply:
Yes, done